File: backend/optimizer/optimizer_api.py
# optimizer_api.py - JSON output version of optimizer for API integration
import json
from network import RailwayNetwork
from scheduler_with_platforms import TrainScheduler

def run_optimizer():
    try:
        # Initialize network (same as main.py)
        network = RailwayNetwork()
        stations = ["A", "B", "C", "D"]
        for s in stations:
            network.add_station(s)
        network.add_track("A", "B", 5, 60)
        network.add_track("B", "C", 10, 80)
        network.add_track("C", "D", 7, 70)
        network.add_track("A", "C", 15, 50)

        # Define trains with more parameters for frontend
        trains = [
            {
                "id": "T1", 
                "name": "Express Train",
                "type": "express", 
                "priority": 1, 
                "start": "A", 
                "end": "D", 
                "departure": 0, 
                "speed": 60,
                "status": "scheduled",
                "delay": 0
            },
            {
                "id": "T2", 
                "name": "Passenger Train",
                "type": "passenger", 
                "priority": 2, 
                "start": "A", 
                "end": "D", 
                "departure": 1, 
                "speed": 50,
                "status": "running",
                "delay": 3
            },
            {
                "id": "T3", 
                "name": "Freight Train",
                "type": "freight", 
                "priority": 3, 
                "start": "B", 
                "end": "D", 
                "departure": 3, 
                "speed": 40,
                "status": "waiting",
                "delay": 0
            },
        ]

        # Schedule trains
        scheduler = TrainScheduler(network, trains, platforms_per_station=2)
        train_schedule = scheduler.schedule_trains()

        # Build response data structure
        response_data = {
            "success": True,
            "timestamp": "2025-09-21T11:30:00Z",
            "network": {
                "stations": [
                    {"id": s, "name": f"Station {s}", "platforms": 2} for s in stations
                ],
                "tracks": [
                    {"from": "A", "to": "B", "length": 5, "speedLimit": 60},
                    {"from": "B", "to": "C", "length": 10, "speedLimit": 80},
                    {"from": "C", "to": "D", "length": 7, "speedLimit": 70},
                    {"from": "A", "to": "C", "length": 15, "speedLimit": 50}
                ]
            },
            "trains": [],
            "optimizationMetrics": {
                "totalDelay": sum(train["delay"] for train in trains),
                "throughput": 85,
                "efficiency": 92,
                "conflicts": 0
            }
        }

        # Add scheduled train data
        for train in trains:
            train_data = {
                "id": train["id"],
                "name": train["name"],
                "type": train["type"],
                "priority": train["priority"],
                "start": train["start"],
                "end": train["end"],
                "departure": train["departure"],
                "speed": train["speed"],
                "status": train["status"],
                "delay": train["delay"],
                "route": [],
                "schedule": {}
            }

            # Add schedule if available
            if train["id"] in train_schedule:
                schedule = train_schedule[train["id"]]
                for station, (arrival, platform) in schedule.items():
                    train_data["schedule"][station] = {
                        "arrival": arrival,
                        "platform": platform
                    }
                # Build route from schedule
                train_data["route"] = [train["start"]] + list(schedule.keys())

            response_data["trains"].append(train_data)

        # Output JSON to stdout
        print(json.dumps(response_data, indent=2))
        return response_data

    except Exception as e:
        error_response = {
            "success": False,
            "error": str(e),
            "timestamp": "2025-09-21T11:30:00Z"
        }
        print(json.dumps(error_response, indent=2))
        return error_response

if __name__ == "__main__":
    run_optimizer()

File: frontend/src/api/optimizerApi.js
// optimizerApi.js - Frontend API service for connecting to backend optimizer
const API_BASE_URL = 'http://localhost:5000/api/optimizer';

class OptimizerAPI {
  constructor() {
    this.baseURL = API_BASE_URL;
  }

  // Fetch network topology and structure
  async getNetworkData() {
    try {
      const response = await fetch(`${this.baseURL}/network`);
      if (!response.ok) {
        throw new Error(`Network request failed: ${response.status}`);
      }
      const data = await response.json();
      return data;
    } catch (error) {
      console.error('Error fetching network data:', error);
      throw error;
    }
  }

  // Fetch train schedules and optimization results
  async getScheduleData() {
    try {
      const response = await fetch(`${this.baseURL}/schedule`);
      if (!response.ok) {
        throw new Error(`Schedule request failed: ${response.status}`);
      }
      const data = await response.json();
      return data;
    } catch (error) {
      console.error('Error fetching schedule data:', error);
      throw error;
    }
  }

  // Run optimization with custom parameters
  async optimizeSchedule(trains, constraints = {}) {
    try {
      const response = await fetch(`${this.baseURL}/optimize`, {
        method: 'POST',
        headers: {
          'Content-Type': 'application/json',
        },
        body: JSON.stringify({ trains, constraints }),
      });
      
      if (!response.ok) {
        throw new Error(`Optimization request failed: ${response.status}`);
      }
      
      const data = await response.json();
      return data;
    } catch (error) {
      console.error('Error running optimization:', error);
      throw error;
    }
  }

  // Get real-time train status updates
  async getTrainStatus(trainId = null) {
    try {
      const url = trainId 
        ? `${this.baseURL}/status/${trainId}` 
        : `${this.baseURL}/status`;
      
      const response = await fetch(url);
      if (!response.ok) {
        throw new Error(`Status request failed: ${response.status}`);
      }
      
      const data = await response.json();
      return data;
    } catch (error) {
      console.error('Error fetching train status:', error);
      throw error;
    }
  }

  // Convert backend data to frontend format
  convertTrainDataToFrontend(backendTrains) {
    return backendTrains.map(train => ({
      id: train.id,
      name: train.name || `Train ${train.id}`,
      number: this.generateTrainNumber(train.type, train.id),
      type: train.type,
      priority: train.priority,
      section: this.mapStationToSection(train.start),
      speed: train.speed,
      destination: this.getDestinationName(train.end),
      status: this.mapStatus(train.status),
      delay: train.delay || 0,
      route: this.mapRouteToSections(train.route),
      statusType: this.getStatusType(train.status),
      departure: train.departure,
      schedule: train.schedule || {},
      platforms: this.extractPlatformInfo(train.schedule)
    }));
  }

  // Convert backend network data to frontend format
  convertNetworkDataToFrontend(networkData) {
    return {
      stations: networkData.stations?.map(station => ({
        id: station.id,
        name: station.name,
        coordinates: station.coordinates || this.getStationCoordinates(station.id),
        platforms: station.platforms || 2,
        status: 'operational'
      })) || [],
      
      tracks: networkData.tracks?.map(track => ({
        id: `${track.from}-${track.to}`,
        from: track.from,
        to: track.to,
        length: track.length,
        speedLimit: track.speedLimit,
        status: track.status || 'free',
        occupiedBy: track.occupiedBy || null
      })) || []
    };
  }

  // Helper methods
  generateTrainNumber(type, id) {
    const prefixes = {
      'express': '123',
      'passenger': '456', 
      'freight': '789'
    };
    const prefix = prefixes[type] || '000';
    const number = id.replace('T', '');
    return `${prefix}0${number}`;
  }

  mapStationToSection(station) {
    const stationToSection = {
      'A': '1R',
      'B': '3L', 
      'C': '5L',
      'D': '7L'
    };
    return stationToSection[station] || '1R';
  }

  getDestinationName(station) {
    const stationNames = {
      'A': 'New Delhi Junction',
      'B': 'Mumbai Central',
      'C': 'Kolkata Howrah', 
      'D': 'Chennai Central'
    };
    return stationNames[station] || `Station ${station}`;
  }

  mapStatus(status) {
    const statusMap = {
      'scheduled': 'Scheduled',
      'running': 'Running',
      'waiting': 'Waiting',
      'delayed': 'Delayed',
      'stopped': 'Stopped'
    };
    return statusMap[status] || 'Unknown';
  }

  getStatusType(status) {
    if (status === 'delayed') return 'delayed';
    if (status === 'stopped' || status === 'waiting') return 'stopped';
    return 'running';
  }

  mapRouteToSections(route) {
    if (!route) return [];
    return route.map(station => this.mapStationToSection(station));
  }

  extractPlatformInfo(schedule) {
    if (!schedule) return {};
    const platforms = {};
    Object.entries(schedule).forEach(([station, info]) => {
      if (info.platform) {
        platforms[station] = info.platform;
      }
    });
    return platforms;
  }

  getStationCoordinates(stationId) {
    const coordinates = {
      'A': { x: 100, y: 200 },
      'B': { x: 300, y: 200 },
      'C': { x: 500, y: 200 },
      'D': { x: 700, y: 200 }
    };
    return coordinates[stationId] || { x: 0, y: 0 };
  }
}

// Export singleton instance
const optimizerAPI = new OptimizerAPI();
export default optimizerAPI;